### ISSUE TITLE
Sprinker and Faucet services

### ISSUE THREAD

**Issue by migabc** [nfarina/homebridge#1707]
Does homebridge already support Sprinkers and Faucets?

Does anybody know what are the new Service and Characteristic UUIDs for these new HomeKit accessories?

I've just checked the source https://github.com/KhaosT/HAP-NodeJS/blob/master/lib/gen/HomeKitTypes.js for "faucet" and "sprinkler" and there is nothing in it, yet.

P.S. Currently my KNX garden sprinkers are configured on knx_config.json as normal Homekit switches and I would like to change the configuration to Sprinkers. 

Thanks,
Miguel

**Comment by KhaosT** [comment]
Not yet, I'll try to add those soon.

**Comment by migabc** [comment]
I got it working :-)

I'm using the Service.Valve for both my normal valve accessory (ValveType=0) and my Sprinker (ValveType=1) accessory

Characteristic.ValveType.GENERIC_VALVE = 0;
Characteristic.ValveType.IRRIGATION = 1;

It's working nicely.
Thx


/**
 * Service "Valve"												-> All in one Service (use this one and set the Characteristic.ValveType to the appropriate value)
 */

Service.Valve = function(displayName, subtype) {
  Service.call(this, displayName, '000000D0-0000-1000-8000-0026BB765291', subtype);

  // Required Characteristics
  this.addCharacteristic(Characteristic.Active);
  this.addCharacteristic(Characteristic.InUse);
  this.addCharacteristic(Characteristic.ValveType);

  // Optional Characteristics
  this.addOptionalCharacteristic(Characteristic.SetDuration);
  this.addOptionalCharacteristic(Characteristic.RemainingDuration);
  this.addOptionalCharacteristic(Characteristic.IsConfigured);
  this.addOptionalCharacteristic(Characteristic.ServiceLabelIndex);
  this.addOptionalCharacteristic(Characteristic.StatusFault);
  this.addOptionalCharacteristic(Characteristic.Name);
};

inherits(Service.Valve, Service);

Service.Valve.UUID = '000000D0-0000-1000-8000-0026BB765291';

**Comment by tomdrumdrum** [comment]
What is the difference between valvetype generic and irrigation on home app? Does it change the icon? I tried to use both and I am not able to have the sprinkler icon...

**Comment by migabc** [comment]
Yes, ValveType characteristic changes the Home app icon:
with ValveType=0 you get the faucet icon
with ValveType=1 you get the sprinkler icon

**Comment by tomdrumdrum** [comment]
Can you please paste your code here. I don’t understand what is wrong with my code. Thank you very much

**Comment by migabc** [comment]
For the Faucet I have:

	

```
/*******************************************************************************************************************
	 * onKNXValueChange is invoked if a Bus value for one of the bound addresses is received
	 * 
	 */
	onKNXValueChange(field, oldValue, knxValue) {
		// Lets set value for HomeKit

		log('INFO: onKNXValueChange(' + field + ", " + oldValue + ", " + knxValue + ")");

		// Homekit is using maxValue: 1, minValue: 0
		// We need to convert knxValue 0 (INACTIVE) to HKvalue 0; and knxValue 1 (ACTIVE) to HKvalue 1
		// The value property of ValveType must be one of the following:
			// Characteristic.ValveType.GENERIC_VALVE = 0;
			// Characteristic.ValveType.IRRIGATION = 1;
			// Characteristic.ValveType.SHOWER_HEAD = 2;
			// Characteristic.ValveType.WATER_FAUCET = 3;
		if (field==="Active") {
			this.myAPI.setValue("Active", 1 - knxValue);
			this.myAPI.setValue("InUse", 1 - knxValue);
			this.myAPI.setValue("ValveType", 0);
		} // if
	} // onBusValueChange
	
	/*******************************************************************************************************************
	 * onHKValueChange is invoked if HomeKit is changing characteristic values
	 * 
	 */
	onHKValueChange(field, oldValue, newValue) {
		// Lets write/send value to KNX bus

		log('INFO: onHKValueChange(' + field + ", " + oldValue + ", " + newValue + ")");

		if (field==="Active") {
			this.myAPI.knxWrite("Active", 1 - newValue, "DPT1"); // send the new knxValue to the KNX bus
		} // if
	} // onHKValueChange
```

**Comment by tomdrumdrum** [comment]
Thank you! It works now!!

**Comment by migabc** [comment]
(y)

**Comment by justinfanok** [comment]
I wish I could get this working on my Rachio. I've been having some issues modifying the existing plugin

**Comment by jvmahon** [comment]
When using Characteristic.RemainingDuration, is any kind of event generated when the duration timer reaches '0'?  If not, how do you monitor for it?  (I assume I could use setTimeout and start a count usng RemainingDuration.getValue() as the time value to setTimeOut when the "InUse"Characteristic is set to 1 and then close the valve when setTimeout expires, but is there a "better" preferred way?).

**Comment by tonesto7** [comment]
@justinfanok When i am finally free i plan on working on the Rachio plugin to fix it for all of the Rachio v3 API

**Comment by justinfanok** [comment]
@tonesto7 if you need a tester, please let me know!!

**Comment by goedh452** [comment]
I'm making progress with with this, but also am a horrible NOOB. Maybe some can help out...
I'm planning on buying OpenSprinkler which can be controller using HTTP commands. Using the Homebridge-HTTP plugin switch would enable me to see it in the Home app. But the icon is wrong :-)

How can I use the correct service and icon? And where to put the code?

**Comment by jvmahon** [comment]
@goedh452 

I implemented simple "valve" support in my plugin for HomeSeer. Maybe  it will help you get started.

See this file: HomeKitDeviceSetup.js:   https://github.com/jvmahon/homebridge-homeseer/blob/master/lib/HomeKitDeviceSetup.js

Lines 59-142 relate to valve control.

Some notes ...

This code is set to control a simple  On / Off valve in HomeSeer. In HomeSeer, the valve is a simple binary switch. The .updateUsingHSReference and .settConfigValves functions are specific to my HomeSeer implementation. You can ignore them.

To start, you will want to add a valve service (line 61) and 3 characteristics, Active, InUse, and ValveType. Lines 66, 71, 75.

Valve control in HomeKit is quite a bit different than a "regular" switch in that its a two part operation. When the user turns the valve "On" or "Off" in the iOS app, the Active characteristic will be changes and you handle that by an associated .on('change' ...) event handler. The event handler should send the commands to your controller to open the valve. Then, once your controller has responded that the command has been accepted and the valve is open, you set  the "InUse" characteristic to 1. Reverse the process to turn the valve off - i.e., send the "off" commands to your controller, then set "InUse" to 0 and "Active" to 0.

If you look at line 67 of my HomeKitDeviceSetup.js file, you'll see an .on('set' ...) line. This is what gets called when the valve is turned "On" in the Home application. You'll place your function here that gets called to react with your valve by sending HTTP.

When your code determines that the valve has opened, then to update "InUse" you'll then run code along the lines of ...

     valveService.getCharacteristic(Characteristic.InUse).updateValue(1) // Valve has opened.
	 

When you're done, then you would execute:

     valveService.getCharacteristic(Characteristic.InUse).updateValue(0) // Valve to be closed
	 
You also have to similarly set the Active characteristic to 0 and send your command to your controller.

As for the ValveType characteristic, its not that important, but you can set it to 0, 1, 2, 3 to change the icon that shows in the Home app. lines 75-76. 

Now here's the tricky part - the iOS Home application also allows you to select a timer and the valve will shut down after the timer expiration. If you want that feature, you have to include the "SetDuration" characteristic and the RemainingDuration characteristic.  You also have to set up the Javascript that actually counts down the time and shuts the valve. That's being handled at lines 79 - 138. Timer management is a bit tricky as you have to remember that if a user changes the valve time while the valve is open, you have to "remove" the old timer and create a new one. Cancelling out an "old" timer is what is being done at line 90.

Oh, and after all this work on setting up timers, you'll find that the iOS HOme application appears to have a timer bug which (hopefully) will be fixed in iOS 11.4.  The bug is that if you have a valve using a timer, and you lock the iPhone screen, then when you unlock it, the Home app will no longer display the correct remaining duration (though it actually does the timing correctly).

And don't use the "faucets" type. There seems to be a bug Hap-NodeJS. See Pull Request  #567: https://github.com/KhaosT/HAP-NodeJS/pull/567

**Comment by goedh452** [comment]
@jvmahon 

Thank you so much for your detailed answer! I'll dive into it and hopefully get is up and running.

**Comment by timcharper** [comment]
@goedh452 any progress on this? I'm interested in adding my opensprinkler to homebridge also

**Comment by goedh452** [comment]
@timcharper I created a http sprinkler plugin which I am using with homebridge. It can be found here: https://github.com/goedh452/homebridge-http-sprinkler

Not sure if it works with opensprinkler. I use it in combination with Domoticz. In Domoticz I created a switch that opens a valve.

**Comment by timcharper** [comment]
Ahh... so you use Home automation to control your sprinkler schedules?

That looks really straightforward. I'll have a look.

**Comment by timcharper** [comment]
Thanks for the example, that helped a lot.

I have started hacking together a plugin. It works!

https://github.com/timcharper/homebridge-opensprinkler

**Comment by goedh452** [comment]
@dkrizic Scheduling is not implemented yet in my plugin. Once that is done, it is a setting in the Home app.

**Comment by timcharper** [comment]
@dkrizic I'm not sure how to do scheduling with homekit. Plugin is very simple right now. Tap once to turn off valves, tap again to turn on for a preset duration. 

I'd love to make it so you can set the duration for how long a sprinkler will run, and fire off an ad-hoc schedule. I'm just not sure how to expose that. Any pointers?

**Comment by timcharper** [comment]
@erokoder to add to what's already been said, I used these two settings to show the difference between the valve is currently a part of an active program (active), and water is currently flowing through the valve (inUse).

**Comment by lmay** [comment]
Any thoughts on how we can enable the RainBird Wifi Module with Homekit. There is python project that connects to the device - but needs to be wrapped in a homebridge plugin. 

https://github.com/jbarrancos/pyrainbird

**Comment by timcharper** [comment]
@lmay I'd port it over to NodeJS, personally. You can use my homekit OpenSprinkler code as inspiration for writing integration from there. https://github.com/timcharper/homebridge-opensprinkler

**Comment by lmay** [comment]
thanks @timcharper - I think a bit out my league... Likely someone will pick this up. Your project looks great. Thanks for posting

**Comment by goedh452** [comment]
@jvmahon
I’ve got the sprinkler running in the home app through the plugin I created (copy pasted from you and others :-)). 

The sprinkler shows up in the home app and works fine. However, do not see any status information in the first tap indicating that the sprinkler is on. Also I mis the “status and messages” option when going to the settings. Do you have the same issue?

**Comment by jvmahon** [comment]
@goedh452. I don't actually use the Sprinkler accessory even though my HomeSeer plugin supports it. I implemented it at the request of another user and he seemed happy with what I had working, so I then stopped. I assume "status" works for my plugin but have never actually checked. So the only advice I can give you here is that you go to the faucet/sprinkler device's "Details" page, select "Status and Notifications" and then make sure "Include in Status" is turned on for that device.  Sorry I can't be of more help here. Let me know having "Include in Status" turned of was the problem -- if it wasn't, I'll go back and have another look and see how it works in my plugin.

**Comment by goedh452** [comment]
@jvmahon
Thank for the quick response. I’ll dive into the documentation and try to sort it out!

**Comment by goedh452** [comment]
@jahon
I misinterpreted your suggestion. I thought you meant the details page of the code, but you meant the details page in the home app. Strange thing is this option does not appear on the details page in the app. That is why I can’t turn it on. It simply isn’t there. I’m running iOS 12 beta so that might also be the problem. 

Is there any easy way for you to add a sprinkler to your home app with your code an check if the option “include in status” is available on the details page? If not I’ll go searching myself. You’ve been such a great help thus far, so don’t spend to much time on this!

**Comment by jvmahon** [comment]
Yep, I meant the details page of the home app. I can try adding a sprinkler, but might not get to it for a few days. I you have another iPhone / iPad (maybe try a friends by "sharing" your home with the friend?) where you can test this on iOS 11.4, that might be a quicker way for you to check the details page on a full-release iOS device to see if there is a "Status" option, rather than looking on the iOS 12 beta.

**Comment by goedh452** [comment]
You’re right. I’ll try that. Thanks!

**Comment by goedh452** [comment]
@jvmahon 

Still can't find out why the "status and notifications" option is missing. In ioS11 this is also the case. I reported it to apple :-).

Another question, do you have any idea what the optional "Characteristic.IsConfigured" is for? It adds a switch to the details page of the sprinkler with which "zones" can be activated, but I do not now what zones are in relation to sprinklers and cannot find out what effect enabling the switch has.

iOS12 beta has a bug btw which makes the switch unusable, but I'm still curious what it is for and if it can be of any use in my plugin.

Thanks!

**Comment by jvmahon** [comment]
Sorry, no idea how the Characteristic:IsConfigured works. I keep hoping Apple will release an updated version of the HomeKit Accessory Protocol. Maybe when iOS 12 is released :(

**Comment by goedh452** [comment]
Ok, thanks for the response!

**Comment by stevemac00** [comment]
@jvmahon 
Also using your detailed post above with an ESP8266/MQTT on a WaterCop. It works fine turning off and on from iOS but when it turns off/on from the WaterCop I can't seem to get the HomeKit to handle the remote turn on/off appropriately. Homekit will just spin forever saying "Running" or "Stopping". Have you seen this and have any ideas?

**Comment by jvmahon** [comment]
@stevvemac00 - My best guess is it has to do with the interaction between the "Active" and "InUse" characteristics. You have to turn them both on, or both off, in the correct order. I figured it out through a lot of trial / error / guessing.  Sorry, its been a while so I don't recall the details of the specific values you need to use, but it should be in my code..

**Comment by goedh452** [comment]
@stevvemac00
I created a http sprinkler plugin and the polling mechanism check whether the sprinkler is on or off. For it check Domoticz. 

The correct order for the characteristics is:
On: set active to 1 and the set inUse to 1
Off: set inUse to 0 and then active to 0

Maybe this helps.

**Comment by tobekas** [comment]
Hi, I want so share some research regarding sprinkler zones. When I saw the screenshot from the Rachio system with multiple sliders in the Home App (https://9to5mac.com/wp-content/uploads/sites/6/2018/09/HomeKit-sprinkler.png), I wanted to find out how you can do this... and found this out:

- The accessory first needs a `Service.IrrigationSystem` (can be seen as superordinate service).
- For every „Zone“, a `Service.Valve` should be added
  - Every valve service must be linked to the irrigation system: `irrigationSvc.addLinkedService(valveSvc1);` etc.
  - Every valve Service needs a `Characteristic.ServiceLabelIndex` with a different index. ValveType should be set to IRRIGATION, but other types will work as well.
- all services (accessoryInfoSvc, irrigationSvc, valveSvc1, valveSvc2, valveSvc3,...) must be exposed.

I‘m not sure about Zone namings. Even If an optional `Characteristics.Name` is defined for each Valve Service, they will be shown as „Zone 1“, „Zone 2“, ... But it‘s possible to change the names from the Home App afterwards.

A short summary for `Active` / `InUse` - Logic:
- Active=0, InUse=0 -> Off
- Active=1, InUse=0 -> Waiting [Starting, Activated but no water flowing (yet)]
- Active=1, InUse=1 -> Running
- Active=0, InUse=1 -> Stopping


I never found this information in any documentation, so I want to share it here...

**Comment by toshibochan** [comment]
I’m using this! 
https://github.com/sfeakes/SprinklerD

**Comment by ebaauw** [comment]
The setup looks similar to the television setup.  They use _Configured Name_ for the user changeable name of the input channel.

_Active_ is like the target state and _InUse_ like the current state.

**Comment by tobekas** [comment]
btw: If the name of any simple Accessory (switch, lightbulb,...) is changed within the Home App, does this reflect to an HAP characteristic, or is this an app specific override?
Naming is kind of tricky with Homekit. There is the Accessory DisplayName, Information Service Name characteristic, Service Name, Optional Name Characteristic of each Service, the new ConfiguredName Characteristic (read+write) for TV inputsSources...

**Comment by ebaauw** [comment]
> If the name of any simple Accessory (switch, lightbulb,...) is changed within the Home App, does this reflect to an HAP characteristic, or is this an app specific override?

Neither.  It’s updated in the HomeKit configuration, shared between apps and, over iCloud, between devices, but it isn’t reported back to the accessory over HAP.  Very much like rooms or automations.

**Comment by SeydX** [comment]
Hey guys, if anyone is interested in faucet service with temperature adjustment

https://github.com/KhaosT/HAP-NodeJS/issues/665

**Comment by aboulfad** [comment]
So any clues how to use `Characteristic.ProgramMode` ? [link](https://github.com/KhaosT/HAP-NodeJS/blob/a42102f5db0388b67f28974ca5fa489933ca1be6/lib/gen/HomeKitTypes.js#L1540-L1554)

In practice, most Irrigation systems have progam mode, where you program watering days, time, duration, ... it seems this characteristic above is related, but can't find any details on how you program or use that characteristic.

**Comment by tobekas** [comment]
Typically the programming of an irrigation system schedule will be done at the device itself or by an external (proprietary) tool or interface. HomeKit framework is not designed to do the programming part, but the device might tell if a program is scheduled or not. Or if a program is scheduled, but at the moment the system is running by manual activation (PROGRAM_SCHEDULED_MANUAL_MODE_). I think this is just a displaying information.
With HomeKit support, you get the overview about all zones (different Service.Valve instances, linked to the main Service.IrrigationSystem) and their Status and you will be able to activate them manually.

If multiple valves/zone are activated, They might be running parallel or in succession. But I think that you can not adjust this via Homekit. But this could be reflected by the Valve characteristics (e.g. Active=ACTIVATED, but InUse=NOT_IN_USE). So most of the logic has to be done by the device itself or within the Homebridge plugin.

**Comment by aboulfad** [comment]
@tobekas , thanks, you confirmed what i guessed above: that characteristic is just a flag and the heavy lifting (scheduling, time remaining, ...) has to be done by an app/tool/script.

**Comment by denschu** [comment]
Can anyobody explain in more detail how this needs to be configured with homebridge-knx? Maybe @migabc?

**Comment by miticolars** [comment]
Hi, i am also interested in how to control a knx relay module (that i use for my irrigation system) with homebridge-knx plugin. @migabc can you explain how to configure the plugin to control the irrigation system with the home app, please? Thank you!

**Comment by tobekas** [comment]
Have you read Apples official HAP protocol specification? In the current free (non-commercial) version, the usage of these services are explained. You have to use linked services.

As an example for a plugin with IrrigationSystem support, you can have a Look at **homebridge-mqttthing**. You can try around with an mqtt tool like MQTT.fx if you want.
